Compression chapter tuning (#2615)
HTTPArchive/almanac.httparchive.org: src/content/en/2021/compression.md

## Introduction

> We think that Internet users' time is valuable, and that they shouldn't have to wait long for a web page to load. The HTTP protocol allows to compress the responses and this decreases the time needed to transfer the content. Even when taking the time to compress and decompress the content into account, this often leads to significant improvement in the user experience, it can reduce [page weight](./page-weight), improve [web performance](./performance) and boost search rankings, so it's an important part of [search engine optimization](./seo).

Author: Done, thanks

> This chapter discusses lossless compression. Lossy and lossless compression used in [media](../2020/media) formats such as images, audio and video are equally if not more important for increasing page speed, but these are not in the scope of this chapter.

Author: Done, and I worded the first part "applied on a HTTP response", this makes the distinction with part of the file format itself clear

## HTTP compression is useful for many types of content

> We recommend HTTP compression for text-based content, such as [HTML](./markup), [CSS](./css), [JavaScript](./javascript), JSON, or SVG, as well as for woff, ttf and ico. Media files such as images that are already compressed do not benefit from HTTP compression since their representation already includes internal compression.

Author: Done, thanks

{{ figure_markup(
  image="compession-methods-by-content-type.png",
  caption="Compression methods for different content types",
  description="A stacked bar chart showing the usage rate of different compression algorithms broken down by the content type. The stacked bars divide up the use of Brotli, Gzip, and no compression. `text/plain` and `text/html` are the only content types that are compressed less than 50% of the time. `application/json` is compressed approximately 68% of the time, `image/svg+xml` approximately 64%. `text/css` and `application/javascript` are each compressed over 85% of the time, and `application/x-javascript` and `text/javascript` are compressed more than 90% of the time.",
  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vQtfyTM9VEweN_Hli3IuxxqU1CRap4V5Q28baEs7aEBResoPRgk9Dwp1m_vdS9lzNlfO8J4hZN7GPT7/pubchart?oid=1292728213&format=interactive",
  sheets_gid="234112121",
  sql_file="compression_by_content_type.sql"
  )
}}

> Compared to the other content types, `text/plain` and `text/html` uses the least amount of compression, with merely 12% and 14% using compression at all. This might be because `text/html` is more often dynamically generated than static content such as JavaScript and CSS.

Author: I added: `However, it is recommended to compress dynamic content as well, and the next sections show how to configure this.`

Because the next section gives the distinction between dynamic and static content, and then further down the section about how to improve compression gives the recommendation of levels to use for dynamic and for static content

## Server settings for HTTP compression

> For HTTP content encoding, the HTTP standard defines the [Accept-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding) request header, with which a HTTP client can announce to the server what content encodings it can handle. The server\'s response can then contain a [Content-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding) header field that specifies which of the encodings was chosen to transform the data in the response body.

Author: Done, thanks

> Both Brotli and Gzip are supported by virtually all browsers. On the server side, most popular servers like NginX and Apache can be configured to use brotli and/or gzip. We want to give some pointers on how to best configure text compression on your server. There are two types of compression settings:

Reviewer: I think the preferred name of the server is `nginx` (and the company is Nginx), though realise some do use NginX.

```suggestion
Both Brotli and Gzip are supported by virtually all browsers. On the server side, most popular servers like nginx and Apache can be configured to use Brotli and/or Gzip. We want to give some pointers on how to best configure text compression on your server. There are two types of compression settings:
```

Author: Done, and also for other instances of nginx that were capitalized in the chapter

Compression level is a parameter given to the encoder adjusting the amount of effort the encoder is applying to find redundancy in the input, to consequently achieve higher compression density.

> Brotli encoding allows compression levels from 0 to 11, while Gzip uses levels 1 to 9. On top of that, the Zopfli compressor provides a higher level of compression for Gzip, indicated `opt` in the graph below. A higher compression level results in slower compression, but does not substantially affect the decompression speed, even making it slightly faster.

Reviewer: I think the mention of Zopfli and Gzip is a bit confusing since they ultimately are "Gzip" files (at least that is my understanding - but you may know better!). I think the suggested wording here is an improvement for that, but not sure if we could add more on this to make it even clearer? Either here or elsewhere in the text? It was one of the things I was going to ask about when reading the original text.

Author: I tried to reword it to: `Higher levels can be achieved for Gzip as well, with a tool such as Zopfli. This is indicated as `opt` in the graph below.`

After having the minimal form of the content you want to transmit, the next step is to ensure compression is enabled. You can verify it is enabled as highlighted in the previous section, and configure your web server if needed.

> When using only Gzip compression (also known as Deflate or Zlib), you can add support for Brotli. In comparison to Gzip, Brotli compresses to <a hreflang="en" href="https://quixdb.github.io/squash-benchmark/">smaller files at the same speed</a>, decompresses at the same speed, and is widely supported: <a hreflang="en" href="https://caniuse.com/?search=brotli">can I use Brotli</a>.

Reviewer: I always prefer to use the real caniuse link, rather than the search one (just in case some other match - ike a JavaScript Brotli API) comes in above the one you want in the search results in the future:

```suggestion
When using only Gzip compression (also known as Deflate or Zlib), you can add support for Brotli. In comparison to Gzip, Brotli compresses to <a hreflang="en" href="https://quixdb.github.io/squash-benchmark/">smaller files at the same speed</a>, decompresses at the same speed, and is widely supported: <a hreflang="en" href="https://caniuse.com/brotli">can I use Brotli</a>.
```

Author: Done here and also for the other caniuse links elsewhere in the chapter

## Conclusion

> The analysis on compression levels reveals that about 6% of Gzip-compressed content uses more advanced compressors such as Zopfli, while a similar \'optimal parsing\' approach is used for 17% of Brotli-compressed content. This indicates that when much slower but slightly more efficient methods are available, a significant amount of users will deploy these methods for their static content.

Reviewer: ```suggestion
The analysis on compression levels reveals that about 6% of Gzip-compressed content uses more advanced compressors such as Zopfli, while a similar "optimal parsing" approach is used for 17% of Brotli-compressed content. This indicates that when much slower but slightly more efficient methods are available, a significant amount of users will deploy these methods for their static content.
```

Author: Done in the next commit, thanks

> Usage of HTTP compression continues to grow, and especially Brotli has increased significantly compared to the [previous year's Almanac](../2020/compression). The number of HTTP responses using any text compression increased by 2%, while Brotli increased by over 4%. Despite the increase, we still see opportunities to use more HTTP compression by tweaking compression settings of servers. You can benefit from taking a closer look at your own website's responses and your server configuration to enable compression where not yet used, and where used to tweak the compression methods towards higher compression levels. Changing the default compression settings in  popular HTTP servers could have a great impact for users.

Reviewer: ```suggestion
Usage of HTTP compression continues to grow, and especially Brotli has increased significantly compared to the [previous year's chapter](../2020/compression). The number of HTTP responses using any text compression increased by 2%, while Brotli increased by over 4%. Despite the increase, we still see opportunities to use more HTTP compression by tweaking compression settings of servers. You can benefit from taking a closer look at your own website's responses and your server configuration to enable compression where not yet used, and where used to tweak the compression methods towards higher compression levels. Changing the default compression settings in popular HTTP servers could have a great impact for users.
```

The second to last sentence links back well to discussion on why HTML is not compressed. So definitely think it would be good to expand upon that as suggested above, and will make this sentence more powerful.

Author: I added "both for dynamic content such as HTML generated on the fly, and static content" to the end of the sentence here as a reminder of this disctinction
